rust/flatbuffers/src/builder.rs

```diff
@@ -30,6 +31,80 @@ use crate::vector::Vector;
 use crate::vtable::{field_index_to_field_offset, VTable};
 use crate::vtable_writer::VTableWriter;
 
+/// Trait to implement custom allocation strategies for [`FlatBufferBuilder`].
+///
+/// An implementation can be used with [`FlatBufferBuilder::new_in`], enabling a custom allocation
+/// strategy for the [`FlatBufferBuilder`].
+pub trait Allocator: DerefMut<Target = [u8]> {
+    /// Grows the buffer, with the old contents being moved to the end.
+    ///
+    /// Returns the amount it has grown by, so the caller can update offsets.
+    fn grow_downwards(&mut self) -> usize;
```

Comment:
I think returning how much the allocator has grown the buffer by, so the caller can update their offsets, is an abstraction leak.

Ideally, instead of using a normal `usize`, `FlatbufferBuilder` should use a `struct ReverseIndex(usize)` that indexes relative to the end of a slice, so the internal offsets won't need to be updated when the slice grows... but practically, the caller can get the `usize` by checking the length before and after `grow_downwards` is called.

Reply:
We want to support an Allocator backed by some fixed size slice, which means `grow_downwards` should express failures (not that we pipe those errors through `FlatbufferBuilder` yet).

I guess the current API can express such failures by saying the buffer grew by  `0usize`, but I think it'll be more idiomatic to have an associated error type and return `Result<(), Self::Error>`. The default allocator can use `core::convert::Infallible` as an error type since it can't happen.